src/code42cli/file_readers.py

```diff
@@ -17,31 +19,44 @@ def read_csv_arg(headers):
 
 
 def read_csv(file, headers):
-    """Helper to read a csv file object into dict rows, automatically removing header row
-    if it exists, and errors if column count doesn't match header list length.
+    """Helper to read a csv file object into a list of dict rows.
+    If CSV has a header row, all items in `headers` arg must be present in CSV or an
+    error is raised. Any extra columns will get filtered out from resulting dicts.
+
+    If no header row is present in CSV, column count must match `headers` arg length or
+    else error is raised.
     """
-    reader = csv.DictReader(file)
-    first_row = next(reader)
-    if all(header in first_row for header in headers):
-        return [
-            {key: value for (key, value) in row.items() if key in headers}
-            for row in [first_row, *list(reader)]
-        ]
-    else:
-        file.seek(0)
-        reader = csv.DictReader(file, fieldnames=headers)
-        first_row = next(reader)
-        if None in first_row or None in first_row.values():
-            raise click.BadParameter(
-                "Expected headers {} not found in {} and column count doesn't match expected size".format(
-                    headers, file.name
-                )
-            )
-        # skip first row if it's the header values
-        if tuple(first_row.keys()) == tuple(first_row.values()):
-            return list(reader)
+    lines = file.readlines()
+    first_line = lines[0].strip().split(",")
+
+    # handle when first row has all of our expected headers
+    if all(field in first_line for field in headers):
+        csv_rows = []
+        for row in csv.DictReader(lines[1:], fieldnames=first_line):
+            # remove any extra columns that aren't in our expected headers
+            for key in list(row.keys()):
+                if key not in headers:
+                    row.pop(key)
+            csv_rows.append(row)
+        if not csv_rows:
+            raise Code42CLIError("CSV contains no data rows.")
+        return csv_rows
+
+    # handle when first row has no expected headers
+    if all(field not in first_line for field in headers):
```

Comment:
`elif`

I get that there is a return, but it's more readable and can prevent future bugs if we forget to change it to `elif`